_posts/2022-05-23-fine-grained-access-control-for-multitenancy.markdown

```diff
@@ -0,0 +1,369 @@
+
+### Verifying tokens
+
+Token verification is done in 3 steps.
+
+1. Verify structure of token
+2. Verify signature
+3. Verify the claims
+
+#### Verify structure of token
+
+Confirm that the token contains three dot separated base64url strings. If the token does not conform to this structure then it is invalid.
+The first string is a header string followed by a payload string and then finally the signature string as shown below.
+
+```
+<Header>.<Payload>.<Signature>
+```
+
+#### Verify signature
+
+##### 1. Decode token
+
+    To validate the JWT signature, the token must first be decoded.
+
+##### 2. Compare local key ID (kid) to public key ID
+
+&nbsp;&nbsp;&nbsp;&nbsp;
+i) Download and store your JWT's corresponding `JWK` (JSON Web Key) using the following url
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
+`https://cognito-idp.{region}.amazonaws.com/{userPoolId}/.well-known/jwks.json`
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
+substitute region and userPoolId with your user pool's region and user pool ID respectively
+
+&nbsp;&nbsp;&nbsp;&nbsp;
+ii) Search the downloaded `jwks.json` for a `kid` that matches the `kid` of your `JWT`
+
+###### Sample jwks.json
+
+```
+        {
+            "keys": [{
+                "kid": "1234example=",
+                "alg": "RS256",
+                "kty": "RSA",
+                "e": "AQAB",
+                "n": "1234567890",
+                "use": "sig"
+            }, {
+                "kid": "5678example=",
+                "alg": "RS256",
+                "kty": "RSA",
+                "e": "AQAB",
+                "n": "987654321",
+                "use": "sig"
+            }]
+        }
+```
+
+##### 3. Compare signature of the issuer to the signature of the tokens
+
+The signature of the issuer is derived from the JWK with a kid that matches the kid of the JWT.
+Each `JWK` contains an `n` parameter that contains the `modulus value` of the `RSA public key`.
+This is the value that'll be used to derive the `issuer's signature`.
+The JWK will need to be `converted to PEM format` before that can happen.
+
+#### Verify the claims
+
+1. Verify that the token is not expired.
+
+2. The aud claim in an ID token and the client_id claim in an access token should match the app client ID that was created in the Amazon Cognito user pool.
+
+3. The issuer (iss) claim should match your user pool. For example, a user pool created in the us-east-1 Region will have the following iss value:`https://cognito-idp.us-east-1.amazonaws.com/<userpoolID>`
+
+4. Check the `token_use` claim.
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; If you are only accepting the access token in your web API operations, its value must be access.
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; If you are only using the ID token, its value must be id.
+
+## Scenario: Multi-tenant purchase tracking microservice
```

Comment:
Burying the lede again.

As a reader I got bored and went elsewhere before I got to here.

If you don't want to split the article, suggest moving this to the top, as the introduction before signposting the need to introduce a bunch of concepts first.

Also add an `<a>` tag to allow the reader to skip introducing the terms if they are already familiar.

Suggest moving this introduction up to the top